src/v/kafka/server/group.cc

```diff
@@ -1793,93 +1775,89 @@ cluster::abort_group_tx_reply make_abort_tx_reply(cluster::tx_errc ec) {
-              "retrying",
-              r.pid,
-              old_pid,
-              ar);
-            co_return make_begin_tx_reply(cluster::tx_errc::stale);
-        }
-    }
+    auto it = _producers.find(r.pid.get_id());
 
-    auto txseq_it = _tx_data.find(r.pid.get_id());
-    if (txseq_it != _tx_data.end()) {
-        if (r.tx_seq != txseq_it->second.tx_seq) {
+    if (it != _producers.end()) {
+        auto& producer = it->second;
+        if (r.pid.get_epoch() < producer.epoch) {
             vlog(
               _ctx_txlog.warn,
-              "can't begin a tx {} with tx_seq {}: a producer id is already "
-              "involved in a tx with tx_seq {}",
+              "begin tx request failed. Producer {} epoch is lower than "
+              "current fence epoch: {}",
               r.pid,
-              r.tx_seq,
-              txseq_it->second.tx_seq);
-            co_return make_begin_tx_reply(
-              cluster::tx_errc::unknown_server_error);
+              producer.epoch);
+            co_return make_begin_tx_reply(cluster::tx_errc::fenced);
+        } else if (r.pid.get_epoch() > producer.epoch) {
+            // there is a fence, it might be that tm_stm failed, forget about
+            // an ongoing transaction, assigned next pid for the same tx.id and
+            // started a new transaction without aborting the previous one.
+            //
+            // at the same time it's possible that it already aborted the old
+            // tx before starting this. do_abort_tx is idempotent so calling it
+            // just in case to proactively abort the tx instead of waiting for
+            // the timeout
+
+            auto old_pid = model::producer_identity{
+              r.pid.get_id(), producer.epoch};
+            auto ar = co_await do_try_abort_old_tx(old_pid);
+            if (ar != cluster::tx_errc::none) {
+                vlog(
+                  _ctx_txlog.warn,
+                  "begin tx request {} failed, can not abort old transaction: "
+                  "{} - {}",
+                  r,
+                  old_pid,
+                  ar);
+                co_return make_begin_tx_reply(cluster::tx_errc::stale);
+            }
         }
-        if (_ongoing_tx_offsets.contains(r.pid)) {
-            vlog(
-              _ctx_txlog.warn,
-              "can't begin a tx {} with tx_seq {}: it was already begun and it "
-              "accepted writes",
-              r.pid,
-              r.tx_seq);
-            co_return make_begin_tx_reply(
-              cluster::tx_errc::unknown_server_error);
+        if (producer.transaction) {
+            auto& producer_tx = *producer.transaction;
+            if (r.tx_seq != producer_tx.tx_seq) {
+                vlog(
+                  _ctx_txlog.warn,
+                  "begin tx request {} failed - produced has already ongoing "
+                  "transaction with highest sequence number: {}",
```

Comment:
not always the "highest sequence number", right? The condition is just `r.tx_seq != producer_tx.tx_seq`